README.md
[![New Relic Community Plus header](https://raw.githubusercontent.com/newrelic/open-source-office/master/examples/categories/images/Community_Plus.png)](https://opensource.newrelic.com/oss-category/#community-plus)

# Kubernetes Webhook Certificate Manager

[![Build Status](https://travis-ci.com/newrelic/k8s-webhook-cert-manager.svg?branch=master)](https://travis-ci.com/newrelic/k8s-webhook-cert-manager)

Script to generate a certificate suitable for use with any Kubernetes Mutating or Validating Webhook.

To be able to execute the script in a Kubernetes cluster, it's released as a Docker image and can be executed, for instance, as a Kubernetes Job. 

This is a detailed list of steps the script is executing:

- Generate a server key.
- If there is any previous CSR (certificate signing request) for this key, it is deleted.
- Generate a CSR for such key.
- The signature of the key is then approved.
- The server's certificate is fetched from the CSR and then encoded.
- A secret of type tls is created with the server certificate and key.
- The k8s extension api server's CA bundle is fetched.
- The mutating webhook configuration for the webhook server is patched with the k8s api server's CA bundle from the previous step. This CA bundle will be used by the k8s extension api server when calling our webhook.

If you wish to learn more about TLS certificates management inside Kubernetes, check out the official documentation for [Managing TLS Certificate in a Cluster](https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/#create-a-certificate-signing-request-object-to-send-to-the-kubernetes-api).

## Usage example

The script expects multiple mandatory arguments. This is an example:

``` sh
./generate_certificate.sh --service ${WEBHOOK_SERVICE_NAME} --webhook
${WEBHOOK_NAME} --secret ${SECRET_NAME} --namespace ${WEBHOOK_NAMESPACE} 
```

## Development setup

This script is designed to run in Kubernetes clusters. For development purposes, we recommend using [Minikube](https://github.com/kubernetes/minikube).

## Support

Should you need assistance with New Relic products, you are in good hands with several support diagnostic tools and support channels.

>This [troubleshooting framework](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) steps you through common troubleshooting questions.

>New Relic offers NRDiag, [a client-side diagnostic utility](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) that automatically detects common problems with New Relic agents. If NRDiag detects a problem, it suggests troubleshooting steps. NRDiag can also automatically attach troubleshooting data to a New Relic Support ticket. Remove this section if it doesn't apply.

If the issue has been confirmed as a bug or is a feature request, file a GitHub issue.

**Support Channels**

* [New Relic Documentation](https://docs.newrelic.com): Comprehensive guidance for using our platform
* [New Relic Community](https://discuss.newrelic.com): The best place to engage in troubleshooting questions
* [New Relic Developer](https://developer.newrelic.com/): Resources for building a custom observability applications
* [New Relic University](https://learn.newrelic.com/): A range of online training for New Relic users of every level
* [New Relic Technical Support](https://support.newrelic.com/) 24/7/365 ticketed support. Read more about our [Technical Support Offerings](https://docs.newrelic.com/docs/licenses/license-information/general-usage-licenses/support-plan).

## Privacy

At New Relic we take your privacy and the security of your information seriously, and are committed to protecting your information. We must emphasize the importance of not sharing personal data in public forums, and ask all users to scrub logs and diagnostic information for sensitive information, whether personal, proprietary, or otherwise.

We define “Personal Data” as any information relating to an identified or identifiable individual, including, for example, your name, phone number, post code or zip code, Device ID, IP address, and email address.

For more information, review [New Relic’s General Data Privacy Notice](https://newrelic.com/termsandconditions/privacy).

## Contribute

We encourage your contributions to improve this project! Keep in mind that when you submit your pull request, you'll need to sign the CLA via the click-through using CLA-Assistant. You only have to sign the CLA one time per project.

If you have any questions, or to execute our corporate CLA (which is required if your contribution is on behalf of a company), drop us an email at opensource@newrelic.com.

**A note about vulnerabilities**

As noted in our [security policy](../../security/policy), New Relic is committed to the privacy and security of our customers and their data. We believe that providing coordinated disclosure by security researchers and engaging with the security community are important means to achieve our security goals.

If you believe you have found a security vulnerability in this project or any of New Relic's products or websites, we welcome and greatly appreciate you reporting it to New Relic through [HackerOne](https://hackerone.com/newrelic).

If you would like to contribute to this project, review [these guidelines](./CONTRIBUTING.md).

To all contributors, we thank you!  Without your contribution, this project would not be what it is today.

## License

This project is licensed under the [Apache 2.0](http://apache.org/licenses/LICENSE-2.0.txt) License.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.5.0
- Add the `--webhook-number` option to specified how many webhook to certify. Defaults to
  1.

## 1.4.0
- Support multiarch images

## 1.3.0
- Update to alpine 3.12.0

## 1.2.1
- Revert to using the full service name for the CN. There is an open issue in
  EKS in which the SAN is not added to the signed certificates, making the
  TLS requests from the apiserver to the webhook fail.
  https://github.com/awslabs/amazon-eks-ami/issues/341
- Validate that the length of the string "${service}.${namespace}.svc", which
  is used for the CN, is not greater than 64 characters as specified in the
  x509 spec.
- Use ca bundle to patch the webhook from the service account secret instead
  of fetching via kubectl.
- Set the number of retries for retrieving the issued certificate to 10 like
  the error message.
- Add the `--webhook-kind` option to specified between
  MutatingWebhookConfiguration or ValidatingWebhookConfiguration. Defaults to
  MutatingWebhookConfiguration.

## 1.2.0

- Use a much shorter common name for the certificate (only the Service's name)
  to avoid problems due to the character limit in CNs.

## 1.1.1

- Updated musl library to avoid security vulnerability: https://app.snyk.io/vuln/SNYK-LINUX-MUSL-458116

## 1.1.0

### Changed

- Container user is now is `1000` instead of `root`

## 1.0.1

### Added

- Better compatibility with Openshift by patching the webhook configuration json with an `add` operation instead of `replace`.

## 1.0.0
- Initial version of the Kubernetes Webhook Certificate Manager.


e2e-tests/tests.sh
#!/usr/bin/env sh
set -e

SECRET_NAME="e2e-test-cert-manager-secret"
WEBHOOK_NAME="e2e-test-cert-manager-webhook"
JOB_NAME="e2e-test-cert-manager-setup"

# shellcheck disable=SC1090
. "$(dirname "$0")/k8s-e2e-bootstraping.sh"

get_pod_name_by_label() {
    pod_name=""
    i=1
    while [ "$i" -ne 10 ]
    do
        pod_name=$(kubectl get pods -l "$1" -o name | sed 's/pod\///g; s/pods\///g')
        if [ "$pod_name" != "" ]; then
            break
        fi
        sleep 1
        i=$((i + 1))
    done
    printf "%s" "$pod_name"
}

wait_for_pod() {
    set +e
    desired_status=${2:-'Running'}
    is_pod_in_desired_status=false
    i=1
    while [ "$i" -ne 30 ]
    do
        pod_status="$(kubectl get pod "$1" -o jsonpath='{.status.phase}')"
        if [ "$pod_status" = "$desired_status" ]; then
            is_pod_in_desired_status=true
            printf "pod %s is %s\n" "$1" "$desired_status"
            break
        fi

        printf "Waiting for pod %s to be %s\n" "$1" "$desired_status"
        sleep 3
        i=$((i + 1))
    done
    if [ $is_pod_in_desired_status = "false" ]; then
        printf "pod %s does not transition to %s within 1 minute 30 seconds\n" "$1" "$desired_status"
        kubectl get pods
        kubectl describe pod "$1"
        kubectl logs "$1"
        exit 1
    fi
    set -e
}

finish() {
    printf "calling cleanup function\n"
    kubectl delete -f manifests/ || true
}

# build container
(
    cd ..
    make build-container-e2e
    cd -
)

trap finish EXIT

kubectl create -f manifests/webhook.yaml
# add imagePullPolicy=Never in the manifest.
awk '/image: / { print; print "        imagePullPolicy: Never"; next }1' manifests/cert-manager-job.yaml | kubectl create -f -

label="job-name=${JOB_NAME}"
job_pod_name=$(get_pod_name_by_label "$label")
if [ "$job_pod_name" = '' ]; then
    printf 'pod with label %s not found\n' "$label"
    kubectl describe job $JOB_NAME
    exit 1
fi
wait_for_pod "$job_pod_name" "Succeeded"

printf 'secret:\n'
kubectl describe secret $SECRET_NAME
printf '\n'

printf 'mutating webhook configurations:\n'
kubectl get mutatingwebhookconfiguration
printf '\n'

caBundle=$(kubectl get mutatingwebhookconfiguration $WEBHOOK_NAME -o jsonpath='{.webhooks[0].clientConfig.caBundle}')
if [ "$caBundle" = '' ]; then
    printf "caBundle not found for %s\n" "$WEBHOOK_NAME"
    printf "webhook object:\n"
    kubectl describe mutatingwebhookconfiguration $WEBHOOK_NAME
    printf "job logs:\n"
    kubectl logs job/$JOB_NAME
else
    printf "caBundle found in %s object\n" "$WEBHOOK_NAME"
    printf "Tests are passing successfully\n\n"
fi


e2e-tests/k8s-e2e-bootstraping.sh
#!/usr/bin/env sh
# This file is a copy of https://github.com/newrelic/k8s-metadata-injection/blob/master/e2e-tests/k8s-e2e-bootstraping.sh

E2E_KUBERNETES_VERSION=${E2E_E2E_KUBERNETES_VERSION:-}
E2E_MINIKUBE_VERSION=${E2E_E2E_MINIKUBE_VERSION:-}
E2E_SETUP_MINIKUBE=${E2E_SETUP_MINIKUBE:-}
E2E_SETUP_KUBECTL=${E2E_SETUP_KUBECTL:-}
E2E_START_MINIKUBE=${E2E_START_MINIKUBE:-}
E2E_MINIKUBE_DRIVER=${E2E_MINIKUBE_DRIVER:-docker}
E2E_SUDO=${E2E_SUDO:-}

setup_minikube() {
    curl -sLo minikube https://storage.googleapis.com/minikube/releases/"$E2E_MINIKUBE_VERSION"/minikube-linux-amd64 \
        && chmod +x minikube \
        && $E2E_SUDO mv minikube /usr/local/bin/
    
    export MINIKUBE_WANTREPORTERRORPROMPT=false
    export MINIKUBE_HOME=$HOME
    export CHANGE_MINIKUBE_NONE_USER=true
}

setup_kubectl() {
    curl -sLo kubectl https://storage.googleapis.com/kubernetes-release/release/"$E2E_KUBERNETES_VERSION"/bin/linux/amd64/kubectl \
        && chmod +x kubectl \
        && $E2E_SUDO mv kubectl /usr/local/bin/
    
    mkdir "$HOME"/.kube || true
    touch "$HOME"/.kube/config
    export KUBECONFIG=$HOME/.kube/config
}

start_minikube() {
    minikube start --driver="$E2E_MINIKUBE_DRIVER" --kubernetes-version="$E2E_KUBERNETES_VERSION" --logtostderr
}

### Bootstraping

cd "$(dirname "$0")"

[ -n "$E2E_SETUP_MINIKUBE" ] && setup_minikube

minikube version

[ -n "$E2E_SETUP_KUBECTL" ] && setup_kubectl

kubectl version

[ -n "$E2E_START_MINIKUBE" ] && start_minikube

minikube update-context

is_kube_running="false"

set +e
# this for loop waits until kubectl can access the api server that Minikube has created
i=1
while [ "$i" -ne 90 ] # timeout for 3 minutes
do
   kubectl get po 1>/dev/null 2>&1
   if [ $? -ne 1 ]; then
      is_kube_running="true"
      break
   fi

   printf "waiting for Kubernetes cluster up\n"
   sleep 2
   i=$((i + 1))
done

if [ $is_kube_running = "false" ]; then
   minikube logs
   printf "Kubernetes did not start within 3 minutes. Something went wrong.\n"
   exit 1
fi
set -e

kubectl config view


generate_certificate.sh
#!/usr/bin/env sh

set -e

# Fully qualified name of the CSR object
csr="certificatesigningrequests.v1beta1.certificates.k8s.io"

usage() {
  cat <<EOF
Generate certificate suitable for use with any Kubernetes Mutating Webhook.
This script uses k8s' CertificateSigningRequest API to a generate a
certificate signed by k8s CA suitable for use with any Kubernetes Mutating Webhook service pod.
This requires permissions to create and approve CSR. See
https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster for
detailed explanation and additional instructions.
The server key/cert k8s CA cert are stored in a k8s secret.
usage: ${0} [OPTIONS]
The following flags are required.
    --service          Service name of webhook.
    --webhook          Webhook config name.
    --namespace        Namespace where webhook service and secret reside.
    --secret           Secret name for CA certificate and server certificate/key pair.
The following flags are optional.
    --webhook-kind     Webhook kind, either MutatingWebhookConfiguration or
                       ValidatingWebhookConfiguration (defaults to MutatingWebhookConfiguration)
    --webhook-number   Specify the number of webhook (defaults to 1)

EOF
  exit 1
}

while [ $# -gt 0 ]; do
  case ${1} in
      --service)
          service="$2"
          shift
          ;;
      --webhook)
          webhook="$2"
          shift
          ;;
      --secret)
          secret="$2"
          shift
          ;;
      --namespace)
          namespace="$2"
          shift
          ;;
      --webhook-kind)
          kind="$2"
          shift
          ;;
      --webhook-number)
          number="$2"
          shift
          ;;
      *)
          usage
          ;;
  esac
  shift
done

[ -z "${service}" ] && echo "ERROR: --service flag is required" && exit 1
[ -z "${webhook}" ] && echo "ERROR: --webhook flag is required" && exit 1
[ -z "${secret}" ] && echo "ERROR: --secret flag is required" && exit 1
[ -z "${namespace}" ] && echo "ERROR: --namespace flag is required" && exit 1

fullServiceDomain="${service}.${namespace}.svc"

# THE CN has a limit of 64 characters. We could remove the namespace and svc
# and rely on the Subject Alternative Name (SAN), but there is a bug in EKS
# that discards the SAN when signing the certificates.
#
# https://github.com/awslabs/amazon-eks-ami/issues/341
if [ ${#fullServiceDomain} -gt 64 ] ; then
  echo "ERROR: common name exceeds the 64 character limit: ${fullServiceDomain}"
  exit 1
fi

if [ ! -x "$(command -v openssl)" ]; then
  echo "ERROR: openssl not found"
  exit 1
fi


tmpdir=$(mktemp -d)
echo "creating certs in tmpdir ${tmpdir} "

cat <<EOF >> "${tmpdir}/csr.conf"
[req]
req_extensions = v3_req
distinguished_name = req_distinguished_name
[req_distinguished_name]
[ v3_req ]
basicConstraints = CA:FALSE
keyUsage = nonRepudiation, digitalSignature, keyEncipherment
extendedKeyUsage = serverAuth
subjectAltName = @alt_names
[alt_names]
DNS.1 = ${service}
DNS.2 = ${service}.${namespace}
DNS.3 = ${fullServiceDomain}
EOF

openssl genrsa -out "${tmpdir}/server-key.pem" 2048
openssl req -new -key "${tmpdir}/server-key.pem" -subj "/CN=${fullServiceDomain}" -out "${tmpdir}/server.csr" -config "${tmpdir}/csr.conf"

csrName=${service}.${namespace}
echo "creating csr: ${csrName} "
set +e

# clean-up any previously created CSR for our service. Ignore errors if not present.
if kubectl get "${csr}/${csrName}"; then
  if kubectl delete "${csr}/${csrName}"; then
    echo "WARN: Previous CSR was found and removed."
  fi
fi

set -e

# create server cert/key CSR and send it to k8s api
cat <<EOF | kubectl create --validate=false -f -
apiVersion: certificates.k8s.io/v1beta1
kind: CertificateSigningRequest
metadata:
  name: ${csrName}
spec:
  groups:
  - system:authenticated
  request: $(base64 < "${tmpdir}/server.csr" | tr -d '\n')
  usages:
  - digital signature
  - key encipherment
  - server auth
EOF

set +e
# verify CSR has been created
while true; do
  if kubectl get "${csr}/${csrName}"; then
      break
  fi
done

set -e

# approve and fetch the signed certificate
kubectl certificate approve "${csr}/${csrName}"

set +e
# verify certificate has been signed
i=1
while [ "$i" -ne 20 ]
do
  serverCert=$(kubectl get "${csr}/${csrName}" -o jsonpath='{.status.certificate}')
  if [ "${serverCert}" != '' ]; then
      break
  fi
  sleep 3
  i=$((i + 1))
done

set -e
if [ "${serverCert}" = '' ]; then
  echo "ERROR: After approving csr ${csrName}, the signed certificate did not appear on the resource. Giving up after 1 minute." >&2
  exit 1
fi

echo "${serverCert}" | openssl base64 -d -A -out "${tmpdir}/server-cert.pem"

# create the secret with CA cert and server cert/key
kubectl create secret tls "${secret}" \
      --key="${tmpdir}/server-key.pem" \
      --cert="${tmpdir}/server-cert.pem" \
      --dry-run -o yaml |
  kubectl -n "${namespace}" apply -f -

caBundle=$(base64 < /run/secrets/kubernetes.io/serviceaccount/ca.crt  | tr -d '\n')

set +e
# Patch the webhook adding the caBundle. It uses an `add` operation to avoid errors in OpenShift because it doesn't set
# a default value of empty string like Kubernetes. Instead, it doesn't create the caBundle key.
# As the webhook is not created yet (the process should be done manually right after this job is created),
# the job will not end until the webhook is patched.
count=0
while true; do
  echo "INFO: Trying to patch webhook adding the caBundle."
  for i in seq 0 $((${number:-1}-1)); do
    if kubectl patch "${kind:-mutatingwebhookconfiguration}" "${webhook}" --type='json' -p "[{'op': 'add', 'path': '/webhooks/${i}/clientConfig/caBundle', 'value':'${caBundle}'}]"; then
      count=$((count + 1))
    fi

    if [ "${count}" = "${number:-1}" ]; then
      break 2
    fi

    echo "INFO: webhook not patched. Retrying in 5s..."
    sleep 5
  done
done
